File: server/src/routes/fsRoute.js
import express from 'express';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
// Project root = repo root (two levels up from server/src/routes).
const PROJECT_ROOT = path.resolve(__dirname, '../../../');

const IGNORE = new Set(['node_modules', '.git', 'dist', 'uploads', 'qdrant']);

const router = express.Router();

// Resolves a user-supplied relative path against PROJECT_ROOT and rejects
// anything that escapes it (../../etc) — the only thing standing between
// "edit this project's files" and "read any file on the machine".
function resolveSafe(relPath) {
  const target = path.resolve(PROJECT_ROOT, '.' + path.sep + (relPath || ''));
  if (!target.startsWith(PROJECT_ROOT)) {
    throw new Error('Path escapes project root');
  }
  return target;
}

function buildTree(dirAbs, relPath = '') {
  const entries = fs.readdirSync(dirAbs, { withFileTypes: true })
    .filter((e) => !IGNORE.has(e.name) && !e.name.startsWith('.'));

  return entries.map((e) => {
    const entryRel = relPath ? `${relPath}/${e.name}` : e.name;
    if (e.isDirectory()) {
      return { name: e.name, path: entryRel, type: 'dir', children: buildTree(path.join(dirAbs, e.name), entryRel) };
    }
    return { name: e.name, path: entryRel, type: 'file' };
  }).sort((a, b) => (a.type === b.type ? a.name.localeCompare(b.name) : a.type === 'dir' ? -1 : 1));
}

router.get('/tree', (req, res) => {
  try {
    const tree = buildTree(PROJECT_ROOT);
    res.json({ name: path.basename(PROJECT_ROOT), path: '', type: 'dir', children: tree });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

router.get('/file', (req, res) => {
  try {
    const target = resolveSafe(req.query.path);
    const content = fs.readFileSync(target, 'utf-8');
    res.json({ path: req.query.path, content });
  } catch (err) {
    res.status(400).json({ error: err.message });
  }
});

router.put('/file', express.json({ limit: '5mb' }), (req, res) => {
  try {
    const { path: relPath, content } = req.body;
    const target = resolveSafe(relPath);
    fs.writeFileSync(target, content ?? '', 'utf-8');
    res.json({ ok: true });
  } catch (err) {
    res.status(400).json({ error: err.message });
  }
});

export default router;

File: server/app/routes/fs_route.py
import os

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

router = APIRouter()

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../"))
IGNORE = {"node_modules", ".git", "dist", "uploads", "qdrant"}


def _resolve_safe(rel_path: str | None) -> str:
    """Resolves a user-supplied relative path against PROJECT_ROOT and rejects
    anything that escapes it — the only thing standing between "edit this
    project's files" and "read any file on the machine"."""
    target = os.path.abspath(os.path.join(PROJECT_ROOT, "." + os.sep + (rel_path or "")))
    if not target.startswith(PROJECT_ROOT):
        raise ValueError("Path escapes project root")
    return target


def _build_tree(dir_abs: str, rel_path: str = "") -> list[dict]:
    entries = [e for e in os.scandir(dir_abs) if e.name not in IGNORE and not e.name.startswith(".")]

    result = []
    for e in entries:
        entry_rel = f"{rel_path}/{e.name}" if rel_path else e.name
        if e.is_dir():
            result.append({"name": e.name, "path": entry_rel, "type": "dir", "children": _build_tree(os.path.join(dir_abs, e.name), entry_rel)})
        else:
            result.append({"name": e.name, "path": entry_rel, "type": "file"})

    result.sort(key=lambda x: (x["type"] != "dir", x["name"]))
    return result


@router.get("/api/fs/tree")
async def get_tree():
    try:
        tree = _build_tree(PROJECT_ROOT)
        return {"name": os.path.basename(PROJECT_ROOT), "path": "", "type": "dir", "children": tree}
    except Exception as err:
        raise HTTPException(status_code=500, detail=str(err))


@router.get("/api/fs/file")
async def get_file(path: str = Query(...)):
    try:
        target = _resolve_safe(path)
        with open(target, "r", encoding="utf-8") as f:
            content = f.read()
        return {"path": path, "content": content}
    except Exception as err:
        raise HTTPException(status_code=400, detail=str(err))


class WriteFileBody(BaseModel):
    path: str
    content: str | None = None


@router.put("/api/fs/file")
async def write_file(body: WriteFileBody):
    try:
        target = _resolve_safe(body.path)
        with open(target, "w", encoding="utf-8") as f:
            f.write(body.content or "")
        return {"ok": True}
    except Exception as err:
        raise HTTPException(status_code=400, detail=str(err))
